### PR TITLE
Enable P2P shuffling by default

### PR DESCRIPTION
**Pull request by fjetter**
This would enable P2P shuffling by default for _most_ shuffle based dataframe workloads (set_index, groupby w/ou split_out, etc.) iff pyarrow is installed.

There are still a couple of cases that are hard coded to tasks because they went through an elaborate evaluation which is something I don't feel entirely comfortable with toggling right now without even testing at least once. See for some discussion over here https://github.com/dask/dask/pull/9826

We haven't received a lot of feedback (see also https://github.com/dask/distributed/discussions/7509) yet but haven't encountered any critical issues during internal validation that I would suggest to flip the switch unless there are major objections (e.g. CI complains very hard).
The discussion issue also outlines a couple of pros/cons of moving to this new algorithm but I believe the benefits outweigh the cost for almost all users.

### CODE REVIEW

File: dask/utils.py [hunk 1]
```diff
@@ -2091,6 +2091,11 @@ def get_default_shuffle_algorithm() -> str:
         from distributed import default_client
 
         default_client()
-        return "tasks"
+        try:
+            import pyarrow  # noqa
+
+            return "p2p"
+        except ImportError:
+            return "tasks"
```
Comment (fjetter) [inline dask/utils.py:2099]: We could add a warning/info log suggesting to install pyarrow if we hit this exception.

**Review by jrbourbeau — COMMENTED** [review]
Thanks @fjetter! I've not thought too deeply about this yet, but will devote some cycles to it. 

FWIW it looks like the `dask/tests/test_distributed.py::test_fused_blockwise_dataframe_merge` failure is genuinely related to the change in default 

Also cc @quasiben who asked me about this a few days ago

**Comment by jrbourbeau** [comment]
> We haven't received a lot of feedback (see also https://github.com/dask/distributed/discussions/7509)

@rjzamora I think you mentioned wanting to try P2P shuffling out at some point. If you have any feedback, it would certainly be welcome. @quasiben are there folks you're aware of that would be interested in trying P2P and sharing feedback on their experience?

**Comment by quasiben** [comment]
Yes, @wence- met up with @fjetter and others and is going to starting experimenting shortly

**Comment by fjetter** [comment]
It seems we're installing a couple of rather old pyarrow versions on some jobs, e.g. python3.8 jobs are installing pyarrow 4. Not sure what that is since we're not pinning anything.

File: dask/utils.py (hunk 1 above)
Comment (hendrikmakait) [reply dask/utils.py]: Should we add a `p2p` extras target that requires `distributed` and a suitable `pyarrow` version?

Comment (fjetter) [reply dask/utils.py]: A p2p extra feels way too specific. I'd rather add this to the `dataframe` target but that'd obviously be a bigger change

**Comment by fjetter** [comment]
Ok, looks like it's just the python 3.8 builds that are using this ancient pyarrow version. I haven't checked the entire matrix but py3.9+ seems to be using pyarrow 11.X so the tests are actually covering the p2p shuffle ;)

**Comment by hendrikmakait** [comment]
> 

This failure is cause by shuffling and task fusion not playing together nicely at the moment due to our reliance on the task name being unchanged. @fjetter: Could you re-run CI after the latest changes to dask/distributed#7578? The error should then pop up again.

File: dask/utils.py [hunk 2]
```diff
@@ -2091,6 +2092,15 @@ def get_default_shuffle_algorithm() -> str:
         from distributed import default_client
 
         default_client()
+        # We might loose annotations if low level fusion is active
+        if not dask.config.get("optimization.fuse.active"):
```
Comment (fjetter) [inline dask/utils.py:2096]: I added this guard to only toggle to p2p if low level fusion is disabled. We are relying on annotations and low level fusion apparently strips them sometimes.

**Comment by fjetter** [comment]
Tests are looking good. Anybody brave enough to ✅ ? @mrocklin maybe? :)

**Comment by mrocklin** [comment]
I don't have enough hands-on experience with this to approve I think, especially given that we're up against a release day.  Unfortunately I'm going to take the cowardly way here and defer to you and @hendrikmakait , who have the appropriate context here to make this decision.

I will say though that, if you both feel good about this, then I encourage you to move forward with it boldly.

File: dask/utils.py [hunk 3]
```diff
@@ -2091,6 +2092,15 @@ def get_default_shuffle_algorithm() -> str:
         from distributed import default_client
 
         default_client()
+        # We might loose annotations if low level fusion is active
```
Comment (wence-) [inline dask/utils.py:2095]:
```suggestion
        # We might lose annotations if low level fusion is active
```

**Review by hendrikmakait — APPROVED** [review]
Thanks @fjetter! The guards seem reasonable to me and should avoid user pains. Given the lack of negative feedback and the strictly positive results we've seen in testing, I'd say we should go ahead and roll this out. @wence-: Any chance you managed to break things in the meantime and would want to veto this?

**Comment by hendrikmakait** [comment]
As discussed offline, we should document the config value users can set to fallback globally and rename it from a top-level `shuffle` to `dataframe.shuffle` (with a deprecation cycle).

File: dask/dask-schema.yaml [hunk 4]
```diff
@@ -46,13 +46,22 @@ properties:
...
+          description: |
+            Compression algorithm used for on disk-shuffling. Partd, the library used
+            for compression supports ZLib, BZ2, and SNAPPY
+
+        algorithm:
+          type:
+          - string
+          - "null"
+          description: |
+            The default shuffle algorithm so choose. Possible values are disk,
```
Comment (hendrikmakait) [inline dask/dask-schema.yaml:63]:
```suggestion
            The default shuffle algorithm to choose. Possible values are disk,
```

**Comment by wence-** [comment]
>  @wence-: Any chance you managed to break things in the meantime and would want to veto this?

I haven't had a chance yet (and for the cases I was going to try "tasks" is already broken, or at least doesn't complete). In dask-cudf we explicitly override the shuffle default, so I think I don't have any veto reasons.

File: dask/dask.yaml [hunk 5]
```diff
@@ -8,7 +8,9 @@ tokenize:
 
 dataframe:
   backend: "pandas"  # Backend dataframe library for input IO and data creation
-  shuffle-compression: null  # compression for on disk-shuffling. Partd supports ZLib, BZ2, SNAPPY
+  shuffle:
+    algorithm: null
+    compression: null  # compression for on disk-shuffling. Partd supports ZLib, BZ2, SNAPPY
```
Comment (hendrikmakait) [inline dask/dask.yaml:13]: If it's only about disk-based shuffling, what about `dataframe.shuffle.disk.compression`?

Comment (fjetter) [reply dask/dask.yaml:13]: I could see this being used elsewhere and think it's fine for a specific algorithm to ignore such a value and suggest to keep it as is

File: dask/utils.py [hunk 6]
```diff
@@ -2085,12 +2086,21 @@ def is_namedtuple_instance(obj: Any) -> bool:
...
 def get_default_shuffle_algorithm() -> str:
-    if d := config.get("shuffle", None):
+    if d := config.get("dataframe.shuffle.algorithm", None):
         return d
     try:
         from distributed import default_client
 
         default_client()
+        # We might lose annotations if low level fusion is active
+        if not dask.config.get("optimization.fuse.active"):
```
Comment (jrbourbeau) [inline dask/utils.py:2096]: I'm a bit confused about why this is needed (not necessarily saying it's wrong, just that I'm lacking context). @hendrikmakait you said 

> This failure is cause by shuffling and task fusion not playing together nicely at the moment due to our reliance on the task name being unchanged

Can you point me to where I can read more about this?

Comment (jrbourbeau) [reply dask/utils.py:2096]: Possibly related, are we accounting for `dask.dataframe` and `dask.array` treating this config option slightly differently? `dask.array` will perform low-level task fusion by default, while `dask.dataframe` won't. 

`dask.array` 

https://github.com/dask/dask/blob/a71c15b15e05c5896821e9c7420dfa22787a0388/dask/array/optimization.py#L52-L55

`dask.dataframe`

https://github.com/dask/dask/blob/a71c15b15e05c5896821e9c7420dfa22787a0388/dask/dataframe/optimize.py#L27-L31

Sorry for being late to the party, I'm still catching up from some recent PTO

**Pull request merged**